### PR TITLE
MM-44756: playbook join/leave, list view and editor view

### PR DESCRIPTION
**Pull request by calebroseland**
#### Summary
CleanShot 2022-07-22 at 04 46 28

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-44756

#### Checklist

- [ ] Telemetry updated
- [ ] Gated by experimental feature flag
- [ ] Unit tests updated

### CODE REVIEW

File: webapp/src/components/backstage/playbook_editor/controls.tsx [hunk 1]
```diff
@@ -288,6 +302,30 @@ export const RunPlaybook = ({playbook}: ControlProps) => {
...
+export const JoinPlaybook = ({playbook: {id: playbookId}, refetch}: ControlProps & {refetch: () => void;}) => {
+    const {formatMessage} = useIntl();
+    const currentUser = useSelector(getCurrentUser);
+    const {join} = usePlaybookMembership(playbookId, currentUser.id);
+    const {setFollowing} = useEditorFollowersMeta(playbookId);
+
+    return (
+        <PrimaryButtonLarger
+            onClick={async () => {
+                join();
```
Comment (gbochora) [inline webapp/src/components/backstage/playbook_editor/controls.tsx:314]: Do we need to wait for the `join` to complete?
```suggestion
                await join();
```

File: webapp/src/components/backstage/playbook_editor/controls.tsx [hunk 2]
```diff
@@ -346,30 +402,65 @@ const TitleMenuImpl = ({playbook, children, className, editTitle, refetch}: Titl
...
-                        <RedText>
-                            <FormattedMessage defaultMessage='Archive playbook'/>
-                        </RedText>
-                    </DropdownMenuItem>
+                {currentUserMember && (
+                    <>
+                        <div className='MenuGroup menu-divider'/>
+                        <DropdownMenuItem
+                            onClick={() => {
+                                leave();
```
Comment (gbochora) [inline webapp/src/components/backstage/playbook_editor/controls.tsx:429]: Same is here:
```suggestion
                                await leave();
```

**Review by gbochora — APPROVED** [review]
Great work! I left a few comments, but nothing blocking. 
Also, there is an issue with leaving an archived Playbook, it fails silently.

https://user-images.githubusercontent.com/4368372/182576081-1b0fc8fe-0aa4-4b7f-ab9c-a5baa9eb8fbe.mp4

File: webapp/src/graphql/generated_types.ts [hunk 3]
```diff
@@ -66,7 +66,9 @@ export {MetricType};
 export type Mutation = {
     __typename?: 'Mutation';
     addMetric: Scalars['String'];
+    addPlaybookMember: Scalars['String'];
```
Comment (lieut-data) [inline webapp/src/graphql/generated_types.ts:69]: Based on the file name / content, I'm assuming this is generated -- should track this in a .gitattributes as per https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github?

File: server/api/graphql_root.go [hunk 4]
```diff
@@ -382,6 +382,66 @@ func (r *RootResolver) UpdateRun(ctx context.Context, args struct {
...
 
+func (r *RootResolver) AddPlaybookMember(ctx context.Context, args struct {
+	PlaybookID string
+	UserID     string
+}) (string, error) {
+	c, err := getContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	userID := c.r.Header.Get("Mattermost-User-ID")
```
Comment (lieut-data) [inline server/api/graphql_root.go:393]: nit: given we unwrap via `getContext`, I wonder if we can add a helper method to get the user id vs. having to duplicate this string everywhere and know to look on the request header.

File: server/api/graphql_root.go [hunk 5]
```diff
@@ -382,6 +382,66 @@ func (r *RootResolver) UpdateRun(ctx context.Context, args struct {
...
+	currentPlaybook, err := c.playbookService.Get(args.PlaybookID)
+	if err != nil {
+		return "", err
+	}
+
+	if currentPlaybook.DeleteAt != 0 {
+		return "", errors.New("archived playbooks can not be modified")
+	}
+
+	if err := c.permissions.PlaybookManageMembers(userID, currentPlaybook); err != nil {
```
Comment (lieut-data) [inline server/api/graphql_root.go:404]: nit: Should we check permissions before any other constraints?

File: server/api/graphql_root.go [hunk 6]
```diff
@@ -382,6 +382,66 @@ func (r *RootResolver) UpdateRun(ctx context.Context, args struct {
...
+	}
+	userID := c.r.Header.Get("Mattermost-User-ID")
+
+	currentPlaybook, err := c.playbookService.Get(args.PlaybookID)
+	if err != nil {
+		return "", err
+	}
+
+	if currentPlaybook.DeleteAt != 0 {
+		return "", errors.New("archived playbooks can not be modified")
```
Comment (lieut-data) [inline server/api/graphql_root.go:431]: Will this mean I can't leave an archived playbook?

**Review by lieut-data — APPROVED** [review]
A handful of comments, but nothing blocking I think.

File: webapp/src/graphql/generated_types.ts (hunk 3 above)
Comment (crspeller) [reply webapp/src/graphql/generated_types.ts:69]: It is generated bu I actually would like to see it in diffs by default.

File: server/api/graphql_root.go (hunk 4 above)
Comment (crspeller) [reply server/api/graphql_root.go:393]: That makes sense. Will do in a separate PR.

File: server/api/graphql_root.go (hunk 5 above)
Comment (crspeller) [reply server/api/graphql_root.go:404]: I fell like expensive checks should go last, but 0/5.

File: server/api/graphql_root.go (hunk 6 above)
Comment (crspeller) [reply server/api/graphql_root.go:431]: Correct. This is the existing behavior.

**Comment by crspeller** [comment]
I believe cypress test failures are unrelated: https://community.mattermost.com/core/pl/t1zkk7h77bnddk476r3gaacfpo

**Pull request merged**